src/metaswitch/clearwater/cluster_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/python
+
+import etcd
+import json
+
+from .constants import *
+
+
+class EtcdSynchronizer(object):
+
+    def __init__(self, plugin, ip):
+        self._fsm = SyncFSM(plugin, ip)
+        self._ip = ip
+        self._client = etcd.Client(ip, 4000)
+        self._key = plugin._key
+        self._index = None
+        self._leaving_flag = False
+
+    def main(self):
+        # Continue looping while the FSM is running.
+        while self._fsm.is_running:
+            # This blocks on changes to the cluster in etcd.
+            cluster_view = self.read_from_etcd()
+            cluster_state = self.calculate_cluster_state(cluster_view)
+
+            # This node can only leave the cluster if the cluster is in a stable
+            # state. Check the leaving flag and the cluster state. If necessary,
+            # set this node to WAITING_TO_LEAVE. Otherwise, kick the FSM.
+            if self._leaving_flag and cluster_state == STABLE:
+                new_state = WAITING_TO_LEAVE
+            else:
+                local_state = self.calculate_local_state(cluster_view)
+                new_state = self._fsm.next(local_state,
+                                           cluster_state,
+                                           cluster_view)
+
+            # If we have a new state, try and write it to etcd.
+            if new_state:
+                updated_cluster_view = self.update_cluster_view(cluster_view,
+                                                                new_state)
+                self.write_to_etcd(updated_cluster_view)
+
+    # This node has been asked to leave the cluster. Check if the cluster is in
+    # a stable state, in which case we can leave. Otherwise, set a flag and
+    # leave at the next available opportunity.
+    def leave_cluster(self):
+        cluster_view = self.read_from_etcd()
+        cluster_state = self.calculate_cluster_state(cluster_view)
+
+        if cluster_state == STABLE:
+            updated_cluster_view = self.update_cluster_view(cluster_view,
+                                                            new_state)
+            self.write_to_etcd(updated_cluster_view)
+        else:
+            self._leaving_flag = True
+
+    # Calculate the state of the cluster based on the state of all the nodes in
+    # the cluster.
+    def calculate_cluster_state(self, cluster_view):
+        node_states = cluster_view.values()
+
+        if all(state == NORMAL for state in node_states):
+            # All nodes in NORMAL state.
+            return STABLE
+        elif all((state == NORMAL or
+                  state == WAITING_TO_JOIN) for state in node_states):
+            # All nodes in NORMAL or WAITING_TO_JOIN state.
```

Comment:
I'm not sure this comment adds much - `all((state == NORMAL or state == WAITING_TO_JOIN) for state in node_states)` is pretty clear as it is.